apps/prairielearn/src/tests/instructorAssessmentGroups.test.ts

```diff
@@ -0,0 +1,154 @@
+    const getResponse = await fetchCheerio(instructorAssessmentGroupsUrl, {});
+    const csrfToken = extractAndSaveCSRFToken({}, getResponse.$, '[name="add-group-form"]');
+
+    const response = await fetchCookie(fetchCheerio)(instructorAssessmentGroupsUrl, {
+      method: 'POST',
+      body: new URLSearchParams({
+        __csrf_token: csrfToken,
+        __action: 'add_group',
+        group_name: 'TestGroup2',
+        // Add first two users to the group
+        uids: users
+          .slice(0, 2)
+          .map((u) => u.uid)
+          .join(','),
+      }),
+    });
+    assert.equal(response.status, 200);
+    assert.lengthOf(
+      response.$('.alert:contains(The following users are already in another group)'),
+      1,
+    );
+    assert.lengthOf(response.$('#usersTable td:contains(TestGroup2)'), 0);
+  });
+
+  step('can create a second group', async () => {
+    const getResponse = await fetchCheerio(instructorAssessmentGroupsUrl, {});
+    const csrfToken = extractAndSaveCSRFToken({}, getResponse.$, '[name="add-group-form"]');
+
+    const response = await fetchCheerio(instructorAssessmentGroupsUrl, {
+      method: 'POST',
+      body: new URLSearchParams({
+        __csrf_token: csrfToken,
+        __action: 'add_group',
+        group_name: 'TestGroup2',
+        // Add second two users to the group
+        uids: users
+          .slice(2, 4)
+          .map((u) => u.uid)
+          .join(','),
+      }),
+    });
+    assert.equal(response.status, 200);
+    assert.lengthOf(response.$('#usersTable td:contains(TestGroup2)'), 1);
+  });
+
+  step('can add a user to an existing group', async () => {
+    const getResponse = await fetchCheerio(instructorAssessmentGroupsUrl, {});
+
+    // The add member form is dynamically rendered on the client, so we need to
+    // grab the CSRF token from somewhere else instead of getting it from the
+    // actual form.
+    const csrfToken = getCSRFToken(getResponse.$);
+
+    const response = await fetchCookie(fetchCheerio)(instructorAssessmentGroupsUrl, {
+      method: 'POST',
+      body: new URLSearchParams({
+        __csrf_token: csrfToken,
+        __action: 'add_member',
+        group_id: '1',
+        // Add final user to the first group
+        add_member_uids: users[4].uid,
+      }),
+    });
+    assert.equal(response.status, 200);
+    assert.lengthOf(response.$(`#usersTable tr:contains(TestGroup):contains(${users[4].uid})`), 1);
+  });
+
+  step('cannot add a user to a group if they are already in another group', async () => {
+    const getResponse = await fetchCheerio(instructorAssessmentGroupsUrl, {});
+
+    // The add member form is dynamically rendered on the client, so we need to
+    // grab the CSRF token from somewhere else instead of getting it from the
+    // actual form.
+    const csrfToken = getCSRFToken(getResponse.$);
+
+    const response = await fetchCookie(fetchCheerio)(instructorAssessmentGroupsUrl, {
```

Comment:
I'm wondering if `fetchCheerio` should automatically use `fetchCookie` internally? When testing with flash messages, I repeatedly forgot to wrap the fetch with `fetchCookie` to make that work.